core/src/main/java/hudson/model/queue/WorkUnitContext.java

```diff
@@ -130,17 +135,21 @@ public void synchronizeEnd(Queue.Executable executable, Throwable problems, long
      *      the member threads will report {@link InterruptedException}.
      */
     public void synchronizeEnd(Executor e, Queue.Executable executable, Throwable problems, long duration) throws InterruptedException {
```

Comment:
Again, would be better to set the future for an exception where an exception has occurred
```suggestion
    @Restricted(NoExternalUse.class)
    public void synchronizeEnd(Executor e, Queue.Executable executable, Throwable problems, long duration) throws InterruptedException {
        WorkUnit wu = e.getCurrentWorkUnit();
        if (wu == null) {
            throw new IllegalStateException("synchronizeStart shall only be called after a WorkUnit has been assigned");
        }
        boolean mainWork = wu.isMainWork();
        try {
            endLatch.synchronize();

            // the main thread will send a notification
            if (mainWork) {
                if (problems == null) {
                    future.set(executable);
                    e.getOwner().taskCompleted(e, task, duration);
                } else {
                    future.set(problems);
                    e.getOwner().taskCompletedWithProblems(e, task, duration, problems);
                }
            }
        } catch (Error | RuntimeException | InterruptedException problem) {
            if (mainWork) {
                future.set(problem);
            }
            throw problem;
        } finally {
            if (mainWork && !future.isDone()) {
                future.set(executable);
            }
        }
    }
```

Reply:
Let's throw a checked exception then so that we ensure it is proper handling in calling methods for the future

Reply:
Which one are you suggesting? just `new InterruptedException(problem)`? It's not an interrupted exception actually.